utils/signed_files.c

```diff
@@ -0,0 +1,173 @@
+  trit_t root[HASH_LENGTH_TRIT];
+  trit_t public_key_trits[HASH_LENGTH_TRIT];
+  trit_t siblings[depth * HASH_LENGTH_TRYTE * RADIX];
+
+  if ((fp = fopen(signature_filename, "r")) == NULL) {
+    return RC_UTILS_FAILED_TO_OPEN_FILE;
+  }
+
+  curl.type = CURL_P_81;
+  init_curl(&curl);
+
+  normalize_hash(digest, normalized_digest);
+  for (int c = 0; c < HASH_LENGTH_TRYTE; ++c) {
+    long_to_trits(normalized_digest[c], &normalized_digest_trits[c * RADIX]);
+  }
+
+  int i;
+  for (i = 0; i < 3 && (read = getline(&line, &len, fp)) > 0; i++) {
+    line[--read] = '\0';
+    if (read != NORMALIZED_FRAGMENT_LENGTH * HASH_LENGTH_TRYTE) {
+      ret = RC_UTILS_INVALID_SIG_FILE;
+      goto done;
+    }
+    trytes_to_trits((tryte_t *)line, sig_trits, read);
+    iss_curl_sig_digest(
+        sig_digests + i * HASH_LENGTH_TRIT,
+        normalized_digest_trits + i * NORMALIZED_FRAGMENT_LENGTH * RADIX,
+        sig_trits, RADIX * read, &curl);
+    curl_reset(&curl);
+  }
+  if (i != 3 || read < 0) {
+    ret = RC_UTILS_INVALID_SIG_FILE;
+    goto done;
+  }
+
+  iss_curl_address(sig_digests, root, sizeof(sig_digests), &curl);
+  if ((read = getline(&line, &len, fp)) > 0) {
+    line[--read] = '\0';
+    if (read != depth * HASH_LENGTH_TRYTE) {
+      ret = RC_UTILS_INVALID_SIG_FILE;
+      goto done;
+    }
+    trytes_to_trits((tryte_t *)line, siblings, read);
+    merkle_root(root, siblings, depth, index, &curl);
+  } else if (read < 0) {
+    ret = RC_UTILS_INVALID_SIG_FILE;
+    goto done;
+  }
+  trytes_to_trits(public_key, public_key_trits, HASH_LENGTH_TRYTE);
+  *valid = (memcmp(public_key_trits, root, HASH_LENGTH_TRIT) == 0);
+
+done:
+  if (fp) {
+    fclose(fp);
+  }
+  return ret;
+}
+
+static retcode_t digest_file(char const *const filename,
+                             flex_trit_t *const digest) {
+  retcode_t ret = RC_OK;
+  FILE *fp = NULL;
+  ssize_t read = 0;
+  char *line = NULL;
+  size_t len = 0;
+  tryte_t *trytes = NULL;
+  trit_t *trits = NULL;
+  trit_t digest_trits[HASH_LENGTH_TRIT];
+  Kerl kerl;
+
+  if ((fp = fopen(filename, "r")) == NULL) {
+    return RC_UTILS_FAILED_TO_OPEN_FILE;
+  }
+
+  init_kerl(&kerl);
+  while ((read = getline(&line, &len, fp)) > 0) {
+    line[--read] = '\0';
+    // 2 trytes by ASCII character
+    if ((trytes = realloc(trytes, read * 2)) == NULL) {
+      ret = RC_UTILS_OOM;
+      goto done;
+    }
+    ascii_to_trytes(line, trytes);
+    // 3 trits by tryte and size needs to be a multiple of HASH_LENGTH_TRIT
+    // (kerl)
+    if ((trits = realloc(trits, HASH_LENGTH_TRIT *
+                                    (((read * 6) / HASH_LENGTH_TRIT) + 1))) ==
+        NULL) {
+      ret = RC_UTILS_OOM;
+      goto done;
+    }
+    memset(trits, 0, HASH_LENGTH_TRIT * (((read * 6) / HASH_LENGTH_TRIT) + 1));
+    trytes_to_trits(trytes, trits, read * 2);
+    kerl_absorb(&kerl, trits,
+                HASH_LENGTH_TRIT * (((read * 6) / HASH_LENGTH_TRIT) + 1));
```

Comment:
This is a memleak, you need to free `line`. Also: `This buffer should be freed by the user program even if getline() failed.`

Reply:
`getline` realloc the `line` if necessary so we just have to free it once, and it's free'd L144. It's also free if `getline` failed but I just noticed that I'm not returning in this case, I'll fix that.

Reply:
The comment I `copied` above contradicts your `It's also free'd if getline failed

Reply:
Sorry my sentence wasn't clear, I meant in our case, we are freeing it, even it getline failed